opengever/maintenance/scripts/migrate_and_deactivate_meetings.py

```diff
@@ -0,0 +1,303 @@
+            parent = aq_parent(meeting_dossier)
+            data = DexterityObjectDataExtractor(meeting_dossier).extract()
+            data = encode_after_json(data)
+            data[BASEDATA_KEY][u'portal_type'] = u'opengever.dossier.businesscasedossier'
+            del data[FIELDDATA_KEY][u'IMeetingDossier']
+            data[FIELDDATA_KEY][u'IBusinessCaseDossier'] = {}
+            data[FIELDDATA_KEY][u'IProtectDossier'] = {}
+            data = decode_for_json(data)
+            dossier = DexterityObjectCreator(data).create_in(parent)
+
+            # Move all content of meeting dossier to normal dossier
+            for obj in meeting_dossier.contentValues():
+                api.content.move(obj, dossier)
+
+            # update reference in meeting
+            meeting = meeting_dossier.get_meeting()
+            meeting.dossier_oguid = Oguid.for_object(dossier)
+
+            # update favorites
+            query = Favorite.query.by_object(meeting_dossier)
+            query.update({'oguid': Oguid.for_object(dossier)})
+
+            # Save state of the meeting dossier on the new dossier
+            migration_annotations = self.get_migration_annotations(dossier)
+            migration_annotations['state'] = api.content.get_state(meeting_dossier)
+
+            # We will delete the dossiers afterwards, as it otherwise seems
+            # to modify meeting_dossiers during iteration.
+            to_delete.append(meeting_dossier)
+
+        for obj in to_delete:
+            # delete meeting_dossier
+            api.content.delete(obj)
+
+    def migrate_agendaitems_to_subdossiers(self):
+        """ We create a subdossier in the meeting dossier for each agendaitem
+        and copy or move its documents into that subdossier.
+        """
+        for meeting in Meeting.query:
+            meeting_dossier = meeting.get_dossier()
+            responsible = IDossier(meeting_dossier).responsible
+            for agendaitem in meeting.agenda_items:
+                # create a subdossier
+                dossier = api.content.create(
+                    type='opengever.dossier.businesscasedossier',
+                    title=agendaitem.get_title(include_number=True, formatted=True),
+                    responsible=responsible,
+                    container=meeting_dossier)
+
+                # Copy documents from submitted proposal into subdossier
+                for doc in agendaitem.get_documents():
+                    self.copy_and_add_to_mapping(doc, dossier, meeting_dossier)
+
+                # Move excerpts
+                for doc in agendaitem.get_excerpt_documents():
+                    self.move_and_add_to_mapping(doc, dossier, meeting_dossier)
+
+                # Move the proposal document
+                document = agendaitem.resolve_document()
+                self.move_and_add_to_mapping(document, dossier, meeting_dossier)
+
+    def set_meeting_dossier_state(self):
+        for meeting in Meeting.query:
```

Comment:
This could also be invoked from `replace_meeting_dossier_with_normal_dossier()` (after setting the state), which would avoid having to find the new "meeting" dossiers by re-querying.

I'm not worried about performance, but the fact that this `Meeting.query` relies on SQLAlchemy returning us the _new_ meeting dossiers, after their reference has been updated, but not been committed yet. In other words, it depends on [dirty reads](https://www.sqlshack.com/dirty-reads-and-the-read-uncommitted-isolation-level/).

I don't think it's a huge deal, but it sort of makes us more reliant on particular isolation levels, and might possibly behave differently on Oracle.

Reply:
I'm not sure that would work, as I would then need to move stuff from the agendaitems into dossiers which might be resolved or inactive, which would lead to permission issues. I'll store the list of dossiers.
So I've reworked the script to iterate over the meetings just once and complete the migration of a single meeting before moving to the next.